### PR TITLE
storage: paranoid buckets

### LINKED ISSUE #173: bucket_ref on replica does not prevent bucket move
Bucket refs are used to prevent bucket movements during an active request. But only master now makes a decision about rebalancing, without consulting replicas. Master should ask replicas if they have any refs.

Maybe it will require prohibition of rebalancing when a replica is not available.

### PR DESCRIPTION
**Pull request by Gerold103**
The patchset makes `_bucket` space validate its changes via `on_replace` triggers.
Closes #173

### CODE REVIEW

**Review by Serpentian — APPROVED** [review]
Thanks for the patch. Cannot see any problems with it. LGTM.

However, AFAICS it doesn't solve the problem properly: we indeed get rid of the data inconsistency during the running request, but we can still lose the data. Let's extend the example you gave in the commit message of the 3rd part:

```
- Node1 is a master, node2 is a replica, bucket1 is ACTIVE on
  both;
- Node1 sends bucket1 to another replicaset;
- The bucket becomes SENT on both nodes;
- Node1 and node2 loose replication. But netbox still works;
- Node1 and node2 drop each other from config and from
  replication;
- Each of them becomes master and deletes the SENT bucket.
- Node2 receives the bucket from another replicaset again. Now it
  is ACTIVE here and not present on node1 at all. It wasn't accessed
  on the node2, so the ref doesn't exists.
- On the other replicaset the bucket becomes SENT and then it is deleted
- Node1 and node2 again get their old config and restore the
  replication.
- Node2 receives removal of the bucket from node1.
- As a result, bucket doesn't exist anywhere
```

Am I right? If so, I propose to create the issue for strict checking of the bucket state transitions, merge this PR and close #173 already. If the problem occurs some day, we'll at least know what is the reason.

**Comment by Gerold103** [comment]
@Serpentian thanks for the review, I created the follow up ticket: https://github.com/tarantool/vshard/issues/377.

@olegrok PTAL in case you are still interested in this project, I could use a second review.

**Review by olegrok — APPROVED** [review]
Thanks. LGTM.

File: test/rebalancer/rebalancer.test.lua [hunk 1]
```diff
@@ -66,8 +66,16 @@ vshard.storage.rebalancer_disable()
 test_run:switch('box_2_a')
 for i = 1, 100 do _bucket:replace{i, vshard.consts.BUCKET.ACTIVE} end
 
+test_run:switch('default')
+util.map_evals(test_run, {REPLICASET_1, REPLICASET_2},                          \
+               [[vshard.storage.internal.is_bucket_protected = false]])
 test_run:switch('box_1_a')
 _bucket:truncate()
+test_run:switch('default')
```
[the remark this answers is not in the document]
Comment (Gerold103) [reply test/rebalancer/rebalancer.test.lua:74]: Ayoooo indeed, me forgot about them vclock waits, shieeeeet. Thanks for noticing, fixed this and a couple of other places - I didn't need to do `map_evals()` on both replicasets. Porting this to luatest isn't possible in a nice way until we do #107.
```Diff
@@ -67,12 +67,13 @@ test_run:switch('box_2_a')
 for i = 1, 100 do _bucket:replace{i, vshard.consts.BUCKET.ACTIVE} end
 
 test_run:switch('default')
-util.map_evals(test_run, {REPLICASET_1, REPLICASET_2},                          \
+util.map_evals(test_run, {REPLICASET_1},                                        \
                [[vshard.storage.internal.is_bucket_protected = false]])
 test_run:switch('box_1_a')
 _bucket:truncate()
 test_run:switch('default')
-util.map_evals(test_run, {REPLICASET_1, REPLICASET_2},                          \
+test_run:wait_lsn('box_1_b', 'box_1_a')
+util.map_evals(test_run, {REPLICASET_1},                                        \
                [[vshard.storage.internal.is_bucket_protected = true]])
 
 test_run:switch('box_1_a')
@@ -100,12 +101,13 @@ vshard.storage.cfg(cfg, util.name_to_uuid.box_1_a)
 for i = 101, 200 do _bucket:replace{i, vshard.consts.BUCKET.ACTIVE} end
 
 test_run:switch('default')
-util.map_evals(test_run, {REPLICASET_1, REPLICASET_2},                          \
+util.map_evals(test_run, {REPLICASET_2},                                        \
                [[vshard.storage.internal.is_bucket_protected = false]])
 test_run:switch('box_2_a')
 _bucket:truncate()
 test_run:switch('default')
-util.map_evals(test_run, {REPLICASET_1, REPLICASET_2},                          \
+test_run:wait_lsn('box_2_b', 'box_2_a')
+util.map_evals(test_run, {REPLICASET_2},                                        \
                [[vshard.storage.internal.is_bucket_protected = true]])
```

[the remark this answers is not in the document]
Comment (sergos) [reply test/rebalancer/rebalancer.test.lua:74]: Thanks, new version is LGTM.

**Review by sergos — APPROVED** [review]

**Pull request merged**